Huawei ac charger support

MalteSchm opened the pull request:
This PR adds support for AC charging using a Huawei R4850G2. It is the pull request for the feature discussed here: https://github.com/helgeerbe/OpenDTU-OnBattery/discussions/122

- Communication is realized via CAN using a MCP2515 that is connected via (H)SPI. This design was chosen to maintain CAN bus connectivity with the Pylontech batteries (which happens at a different CAN bus speed) and to avoid interfering with the NRF24L01 which uses the (V)SPI bus. 
- The unit can be controlled via MQTT or the Webinterface
- Information is displayed on the Webinterface

Adminius commented:
Project introduction:
https://www.akkudoktor.net/forum/stell-dein-batterie-powerwall-projekt-vor/ac-dc-speicherloesung-mit-victron-mppt-pylontech-hoymiles-huawei-und-opendtu-onbattery/

berni2288 commented:
Very nice, thanks for this great PR!
Maybe I'll also buy myself a Huawei charger to benefit from hourly energy bills :)

I haven't looked much into it yet, just one thing that came into my mind:
The "Limit setzen, Can BUS verbunden" - button - It's not quite clear why there are 2 buttons. The ESP32 should know itself whether there is a CAN BUS connected or not? Maybe it's needed, but I think it should be more understandable for a dummy user  what this is good for, without having to read the (non-existing) documentation for the buttons.

helgeerbe commented on src/InverterSettings.cpp:
```diff
@@ -19,7 +19,7 @@ void InverterSettingsClass::init()
     // Initialize inverter communication
     MessageOutput.print(F("Initialize Hoymiles interface... "));
     if (PinMapping.isValidNrf24Config()) {
-        SPIClass* spiClass = new SPIClass(HSPI);
```
This affects the original OpenDTU. Is this change necessary, that your PR will work? Otherwise I would prefer, that you create a separate PR against the original OpenDTU. So we could ensure, to share the same code base.

MalteSchm replied:
O.k. will do. The code will also work without that change (obviously there is a related change needed in Huawei_can.cpp if this change here is removed).
The reason why I updated this is that the original implementation uses the HSPI interface but with the default Pins of the VSPI controller. This seems to be an ugly inconsistency.

MalteSchm replied:
Ok PR is updated with the inverter SPI interface reverted to the original. I created a PR against openDTU. Lets see what comes out of that

MalteSchm commented:
> I haven't looked much into it yet, just one thing that came into my mind: The "Limit setzen, Can BUS verbunden" - button - It's not quite clear why there are 2 buttons. The ESP32 should know itself whether there is a CAN BUS connected or not? Maybe it's needed, but I think it should be more understandable for a dummy user what this is good for, without having to read the (non-existing) documentation for the buttons.

The Huawei unit has two different Voltage / Ampere limits that can be configured. One limit applies when the unit detects an established CAN connection (online). The other limit is used when there is no CAN link. 

Obviously you cannot set the offline limit (i.e. when the CAN bus is disconnected) with a disconnected CAN bus :-)

MalteSchm commented:
I think I have fixes for the open issues but will test based on the
development brach first before creating another PR

Am Mo., 27. März 2023 um 23:03 Uhr schrieb helgeerbe <
***@***.***>:

> I pulled it into development. Can you test it please!
>
> Got lint errors in actions.
> include/Huawei_can.h:78: Use int16/int64/etc, rather than the C type long
> [runtime/int] [4]
> include/Huawei_can.h:84: Use int16/int64/etc, rather than the C type long
> [runtime/int] [4]
> include/Huawei_can.h:85: Use int16/int64/etc, rather than the C type long
> [runtime/int] [4]
> src/Huawei_can.cpp:46: Use int16/int64/etc, rather than the C type long
> [runtime/int] [4]
> src/Huawei_can.cpp:14
> <https://github.com/helgeerbe/OpenDTU-OnBattery/actions/runs/4536214298/jobs/7992630578#step:5:15>5:
> Use int16/int64/etc, rather than the C type long [runtime/int] [4]
>
> PinMapping in DeviceManager is missing
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/helgeerbe/OpenDTU-OnBattery/pull/144#issuecomment-1485858333>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AVCHAS772I2MRW3LDR65QZDW6H6CLANCNFSM6AAAAAAWIDI6YI>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Robert Hirsch und Malte Schmidt
Adverrun Drives
Fruehlingstr. 41
90537 Feucht
***@***.***

MalteSchm commented:
BTW: I submitted a PR to openDTU to change the SPI interface but there are some side effects with newer chipsets that only offer a single SPI interface. 
This may require moving the Huawei PSU and the NRF interface to a single SPI bus in the future. 
Right now I have not seen them being sold so I think this is more theoretical right now.

MalteSchm commented:
... and thanks for merging

helgeerbe commented:
I have actual problems with the development branch. Systems ui is slow and and sometimes ununresponsive. Got also memory issues due to memory fragmentation. But can't work on it until next week.

Maybe you can debug a little. Maybe you find something

Adminius commented:
> Systems ui is slow and and sometimes ununresponsive

maybe it's also realted to my MQTT issue?

helgeerbe commented:
Could be. Install for testing the master branch. I merged the latest changes from original opendtu. Development branch contains pr for the AC charger. 

But with both branches I don't see any Mqtt connection problems. On development system stops publishing and no livedata is displayed. Problem is that for livedata the JasonDocument becomes too large. System does not get a free memory block of this size. Reboot solve this temporarily.

MalteSchm commented:
Ah I had some MQTT problems as well and tested various things. There was a
point in time where I suspected high flash usage could be the issue (I was
at 99.7%) and I merged the latest openDTU changes with the partition
update. This solved the MQTT problems.

The exact issue is unknown but I suspected a memory overrun as things
suddenly broke when I added MQTT messages one after another.

It was kind of odd that this solved it but I also did not see a point in
investigating this further.



Am Mi., 29. März 2023 um 13:33 Uhr schrieb helgeerbe <
***@***.***>:

> Could be. Install for testing the master branch. I merged the latest
> changes from original opendtu. Development branch contains pr for the AC
> charger.
>
> But with both branches I don't see any Mqtt connection problems. On
> development system stops publishing and no livedata is displayed. Problem
> is that for livedata the JasonDocument becomes too large. System does not
> get a free memory block of this size. Reboot solve this temporarily.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/helgeerbe/OpenDTU-OnBattery/pull/144#issuecomment-1488431349>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AVCHASY4Y4HKQHGJAAIHYLDW6QM2HANCNFSM6AAAAAAWIDI6YI>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Robert Hirsch und Malte Schmidt
Adverrun Drives
Fruehlingstr. 41
90537 Feucht
***@***.***

MalteSchm commented:
I had a look and I believe the issue is in PowerMeter.cpp.
config.PowerMeter.... has not been set on the UI and has a stringlen of 0.
I suppose that MqttSettings.subscribe doesn't like this. Checking for
PowerMeter_Enabled solved this issue.

I combined PR is here: hhttps://
github.com/helgeerbe/OpenDTU-OnBattery/pull/151
I also fixed some issues reported regarding the AC charger. Sorry for
putting all of this in a single PR. I think the changes are simple to
understand though.

Am Mi., 29. März 2023 um 15:08 Uhr schrieb Malte Schmidt <
***@***.***>:

> Ah I had some MQTT problems as well and tested various things. There was a
> point in time where I suspected high flash usage could be the issue (I was
> at 99.7%) and I merged the latest openDTU changes with the partition
> update. This solved the MQTT problems.
>
> The exact issue is unknown but I suspected a memory overrun as things
> suddenly broke when I added MQTT messages one after another.
>
> It was kind of odd that this solved it but I also did not see a point in
> investigating this further.
>
>
>
> Am Mi., 29. März 2023 um 13:33 Uhr schrieb helgeerbe <
> ***@***.***>:
>
>> Could be. Install for testing the master branch. I merged the latest
>> changes from original opendtu. Development branch contains pr for the AC
>> charger.
>>
>> But with both branches I don't see any Mqtt connection problems. On
>> development system stops publishing and no livedata is displayed. Problem
>> is that for livedata the JasonDocument becomes too large. System does not
>> get a free memory block of this size. Reboot solve this temporarily.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/helgeerbe/OpenDTU-OnBattery/pull/144#issuecomment-1488431349>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AVCHASY4Y4HKQHGJAAIHYLDW6QM2HANCNFSM6AAAAAAWIDI6YI>
>> .
>> You are receiving this because you authored the thread.Message ID:
>> ***@***.***>
>>
>
>
> --
> Robert Hirsch und Malte Schmidt
> Adverrun Drives
> Fruehlingstr. 41
> 90537 Feucht
> ***@***.***
>
>

-- 
Robert Hirsch und Malte Schmidt
Adverrun Drives
Fruehlingstr. 41
90537 Feucht
***@***.***
